apps/src/dance/ai/DanceAiModal.tsx

```diff
@@ -148,6 +148,58 @@ const DanceAiModal: React.FunctionComponent = () => {
     (state: {dance: DanceState}) => state.dance.aiOutput
   );
 
+  const getGeneratedEffect = useCallback((step: number) => {
+    if (step < BAD_GENERATED_RESULTS_COUNT) {
+      return generatedEffects.current.badEffects[step];
+    } else if (generatedEffects.current.goodEffect) {
+      return generatedEffects.current.goodEffect;
+    } else {
+      return undefined;
+    }
+  }, []);
+
+  const getScores = useCallback(
+    (useInputs: string[], step: number) => {
+      const effect = getGeneratedEffect(step);
+      if (effect) {
+        return getGeneratedEffectScores(useInputs, effect);
+      }
+
+      return [0, 0, 0];
+    },
+    [getGeneratedEffect]
+  );
+
+  // Calculates the minimum individual score
+  // (ie, a SINGLE emoji association with a foreground/background palette combination),
+  // and a maximum total score
+  // (ie, the sum of ALL selected emoji's associations with a foreground/background palette combination).
+  // Used to normalize and scale the data for easier differentiation between results by the user.
+  const calculateMinMax = useCallback(
+    (useInputs: string[]) => {
+      // The minimum individual score is selected across all generated effects (bad and good).
+      const minIndividualScore = Array.from(
+        Array(BAD_GENERATED_RESULTS_COUNT + 1).keys()
+      ).reduce((accumulator: number, currentValue: number) => {
+        const scores = getScores(useInputs, currentValue);
+        const min = Math.min(...scores);
```

Comment:
@bencodeorg, why did you need to make a deep copy of `scores` here? Just curious.

Reply:
I don't think there's anything copying here— `Math.min` doesn't want an array as its arguments, so this is spreading out the array into a list of args into it.